File: Luogu/ProblemList/Math/Group/P2561.py
def power(a: int, b: int) -> int:
    """快速幂，等价于 C++ 里的模板 power"""
    res = 1
    while b:
        if b & 1:
            res *= a
        a *= a
        b //= 2
    return res


def solve():
    n = int(input().strip())

    # 预计算 f 和 g 数组，长度开到 21 就够了
    g = [0] * (n + 1)
    g[0] = 0
    g[1] = 1
    for i in range(2, n):
        if i % 2 == 0:
            g[i] = g[i - 1] + i // 2
        else:
            g[i] = g[i - 1] + i // 2 + 1
    N = n * (n + 1) // 2
    ans = 0
    ans += power(2, N)
    ans += 2 * power(2, (N + 2) // 3)
    ans += 3 * power(2, g[n])
    ans //= 6
    print(ans)


if __name__ == '__main__':
    tt = 1
    # tt = int(input().strip())
    for _ in range(tt):
        solve()

File: Luogu/ProblemList/Math/Group/P2561.cpp
#include<bits/stdc++.h>
using i64 = long long;
using ll = long long;
using uint = unsigned int;
using ull = unsigned long long;

using namespace std;

template<class T>
constexpr T power(T a, i64 b) {
    T res = 1;
    for (; b; b /= 2, a *= a) {
        if (b % 2) {
            res *= a;
        }
    }
    return res;
}

void solve()
{
    int n;
    cin>>n;
    int ans = 0;
    int N = n * (n + 1) / 2;
    vector<int>f(21),g(21);
    f[0] = g[0] = 0;
    f[1] = g[1] = 1;
    for(int i = 2;i <= 20;i++){
        if(i % 2 == 0){
            f[i] = f[i / 2 - 1] + g[i / 2];
            g[i] = g[i - 1] + i / 2;
        }
        else{
            f[i] = f[i / 2 + 1] + g[i / 2];
            g[i] = g[i - 1] + i / 2 + 1;
        }
    }
    ans += power(2, N);
    ans += 2 * power(2, (N + 2) / 3);
    ans += 6 * power(2, g[n]);
    ans += 3 * power(2, g[n]);
    ans /= 12;
    cout<<ans<<endl;
    return;
}

int main()
{
    std::ios::sync_with_stdio(0);
    std::cin.tie(0);
    int tt = 1;
    //cin >> tt;
    while(tt--){
        solve();
    }
    return 0;
}